### PR TITLE
:memo: Aiven integration

### PR DESCRIPTION
**Pull request by vcampitelli**
Creating Aiven SSO integration guide.

This depends on [this script](https://github.com/FusionAuth/fusionauth-example-scripts/pull/1) to issue X.509 V3 certificates, as our Key Master only generates V1 and Aiven doesn't support it. I'm waiting for that PR to be merged so I can update this article with the final script path.

### CODE REVIEW

**Comment by vcampitelli** [comment]
Updating article with Mac screenshots

**Review by ColinFrick — COMMENTED** [review]
After following the guide I can't login with FusionAuth. I receive following error in the docker log:
```log
integrate-fusionauth-fusionauth-1  | 2023-05-23 11:28:53.761 AM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
integrate-fusionauth-fusionauth-1  | java.lang.NullPointerException: null
integrate-fusionauth-fusionauth-1  | 	at java.base/java.util.Objects.requireNonNull(Objects.java:208)
integrate-fusionauth-fusionauth-1  | 	at io.fusionauth.pem.PEMDecoder.decode(PEMDecoder.java:106)
integrate-fusionauth-fusionauth-1  | 	at io.fusionauth.pem.domain.PEM.decode(PEM.java:111)
integrate-fusionauth-fusionauth-1  | 	at io.fusionauth.api.service.samlv2.DefaultSAMLv2ProviderService.buildResponse(DefaultSAMLv2ProviderService.java:171)
integrate-fusionauth-fusionauth-1  | 	at io.fusionauth.app.action.samlv2.CallbackAction.get(CallbackAction.java:129)
integrate-fusionauth-fusionauth-1  | 	at jdk.internal.reflect.GeneratedMethodAccessor111.invoke(Unknown Source)
integrate-fusionauth-fusionauth-1  | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
integrate-fusionauth-fusionauth-1  | 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.util.ReflectionUtils.invoke(ReflectionUtils.java:414)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.execute(DefaultActionInvocationWorkflow.java:77)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.perform(DefaultActionInvocationWorkflow.java:60)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.message.DefaultMessageWorkflow.perform(DefaultMessageWorkflow.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.validation.DefaultValidationWorkflow.perform(DefaultValidationWorkflow.java:45)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.security.DefaultSecurityWorkflow.perform(DefaultSecurityWorkflow.java:60)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.parameter.DefaultPostParameterWorkflow.perform(DefaultPostParameterWorkflow.java:49)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.content.DefaultContentWorkflow.perform(DefaultContentWorkflow.java:74)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.parameter.DefaultParameterWorkflow.perform(DefaultParameterWorkflow.java:58)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.parameter.DefaultURIParameterWorkflow.perform(DefaultURIParameterWorkflow.java:92)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.scope.DefaultScopeRetrievalWorkflow.perform(DefaultScopeRetrievalWorkflow.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.action.DefaultActionMappingWorkflow.perform(DefaultActionMappingWorkflow.java:113)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.security.DefaultSavedRequestWorkflow.perform(DefaultSavedRequestWorkflow.java:65)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.cors.CORSRequestWorkflow.perform(CORSRequestWorkflow.java:65)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:50)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.workflow.DefaultMVCWorkflow.perform(DefaultMVCWorkflow.java:108)
integrate-fusionauth-fusionauth-1  | 	at org.primeframework.mvc.PrimeMVCRequestHandler.handle(PrimeMVCRequestHandler.java:72)
integrate-fusionauth-fusionauth-1  | 	at io.fusionauth.http.server.HTTPWorker.run(HTTPWorker.java:50)
integrate-fusionauth-fusionauth-1  | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
integrate-fusionauth-fusionauth-1  | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
integrate-fusionauth-fusionauth-1  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
integrate-fusionauth-fusionauth-1  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
integrate-fusionauth-fusionauth-1  | 	at java.base/java.lang.Thread.run(Thread.java:833)
```

Any advice to fix this?

**Comment by vcampitelli** [comment]
@ColinFrick hmm which FusionAuth version are you running? I've tested it a lot against 1.43.x and 1.45.x and can't see any errors. Can you also send me the generated certificate from the Key Master?

**Comment by mooreds** [comment]
Is this blocked on the fix for 2308, @vcampitelli @ColinFrick , or can we get by with saying "this won't work with docker, please check out issue 2308 for progress on fixing it". I'm comfortable shipping with the latter if you all are.

**Comment by vcampitelli** [comment]
I've added the disclaimer @mooreds

File: site/assets/img/docs/samlv2/aiven/create-api-key.png
Comment (mooreds) [inline site/assets/img/docs/samlv2/aiven/create-api-key.png:1]: Please crop this.

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+Aiven is a cloud-based Database-as-a-Service platform. This document covers the configuration necessary to get it working with FusionAuth as the Identity Provider using SAML v2, allowing users to log in to your Aiven account using their FusionAuth credentials.
+
+=== Prerequisites
+
+This document assumes you have a running instance of FusionAuth and a working Aiven application with a project already created. You will also need admin accounts for both to configure them correctly.
+
+[WARNING]
+====
+This does not work well if you are running FusionAuth on Docker. Please check out https://github.com/FusionAuth/fusionauth-issues/issues/2308[this issue] for progress on fixing it.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:18]:
```suggestion
This does not work if you are running FusionAuth on Docker. Please check out https://github.com/FusionAuth/fusionauth-issues/issues/2308[this issue] for progress on fixing this problem.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+image::samlv2/aiven/add-team-to-project.png[Adding team to a project.,width=1200]
+
+=== Configure SSO
+
+Click on [breadcrumb]#Authentication# in the sidebar and then on [uielement]#Add authentication method# to create one. Name it `FusionAuth`, select `SAML` as the [field]#Method Type# and choose the `Developers` team you just created in [field]#Team to autojoin on signup#. Click [uielement]#Add method# to save.
+
+image::samlv2/aiven/add-authentication-method.png[Adding an authentication method.,width=1200]
+
+In the next screen, write down both [field]#Metadata URL# and [field]#ACS URL# values and click [uielement]#Finish#.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:37]:
```suggestion
In the next screen, record both the [field]#Metadata URL# and [field]#ACS URL# values and click [uielement]#Finish#. You'll use these values later.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+== Configure FusionAuth
+
+=== Generating Certificate
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:41]:
```suggestion
=== Generating the Certificate
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+You need to generate a custom RSA certificate and upload it to your FusionAuth instance's Key Master. To do so, there's a script in our https://github.com/FusionAuth/fusionauth-example-scripts[Example scripts GitHub repository] that can do all of this for you.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:43]:
```suggestion
You need to generate a custom RSA certificate and upload it to your FusionAuth instance's Key Master. To do so you can use a script in our https://github.com/FusionAuth/fusionauth-example-scripts[example scripts GitHub repository].
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+Before running it, you need to create an API Key in your FusionAuth instance. Browse to [breadcrumb]#Settings -> API Keys# and click on icon:plus[role=ui-button green,type=fas]. Set the [field]#Description# as `Certificate generator` and locate `/api/key/import` in the [uielement]#Endpoints# section. Click the `POST` switch to enable it and confirm by clicking on icon:save[role=ui-button blue,type=fas].
+
+image::samlv2/aiven/create-api-key.png[Creating API Key.,width=1200]
+
+After being redirected back to the [breadcrumb]#API Keys# page, click on the `Key` for the created item to reveal its actual value and copy it. You'll have to provide this value to the script soon.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:49]:
```suggestion
After being redirected back to the [breadcrumb]#API Keys# page, click on the `Key` field to reveal its actual value. Then copy it. You'll have to provide this value to the script below.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+image::samlv2/aiven/grab-api-key.png[Grabbing API Key.,width=1200,role=bottom-cropped]
+
+Now that your application is ready, clone our https://github.com/FusionAuth/fusionauth-example-scripts[Example scripts GitHub repository] and execute the script located in `v3-certificate`.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:53]: You could also remote include the file and have them copy paste it.

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+[source,shell]
+.Example response for the script
+----
+Your FusionAuth instance URL (with scheme): http://localhost:9011
+API Key with /api/key/import endpoint:      Dy9bphElA3L3_ayW86T5KvrZkyK1Gj5EDV_2NKU5Ypg
+Name for the generated key:                 Aiven key
+----
+
+After finishing, you'll have a certificate in the Key Master in your FusionAuth instance ready to be used. The script will also print the generated certificate, so make sure to write it down.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:72]:
```suggestion
After finishing, you'll have a certificate in the Key Master in your FusionAuth instance ready to be used. The script will also print the generated certificate. Copy that as well, as you'll need it later.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+=== Creating Application
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:74]:
```suggestion
=== Creating the Application
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+|Aiven field|FusionAuth field
+
+|[field]#Metadata URL#
+|[field]#Issuer#
+
+|[field]#ACS URL#
+|[field]#Authorized redirect URLs#
+|===
+
+Scroll down to the [uielement]#Authentication response# section and change the [field]#Signing key# to the `Aiven key` you created above. Click on icon:save[role=ui-button blue,type=fas] to save your application.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:90]:
```suggestion
Scroll down to the [uielement]#Authentication response# section and change the [field]#Signing key# to the `Aiven key` you imported above. Click on icon:save[role=ui-button blue,type=fas] to save your application.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+When redirected to the [breadcrumb]#Applications# page, view your application details by clicking on icon:search[role=ui-button green,type=fas]. In [uielement]#SAML v2 Integration details#, write down both [field]#Entity Id# and [field]#Login URL# fields.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:92]:
```suggestion
When redirected to the [breadcrumb]#Applications# page, view your application details by clicking on icon:search[role=ui-button green,type=fas]. In [uielement]#SAML v2 Integration details#, record both the [field]#Entity Id# and [field]#Login URL# fields.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+Go back to your https://console.aiven.io/[Aiven console] and click [breadcrumb]#Admin# and then on [breadcrumb]#Authentication# in the sidebar to list the authentication methods again. Click on `FusionAuth`, which will still have a `PENDING CONFIGURATION` state until you finish setting it up.
+
+Scroll down to the [uielement]#SAML config# section and click the [uielement]#Edit# button and fill in the information below.
+
+* [field]#SAML IDP Url#: the [field]#Login URL# from your FusionAuth application.
+* [field]#SAML Entity ID#: the [field]#Entity Id# from your FusionAuth application.
+* [field]#SAML Signature Algorithm#: you can leave this blank.
+* [field]#SAML Digest Algorithm#: you can leave this blank.
+* [field]#SAML Field Mapping#: you can leave this blank.
+* [field]#SAML Certificate#: paste the certificate you've copied from the `Generating certificate` two steps above.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:105]:
```suggestion
* [field]#SAML Certificate#: paste the certificate you've copied from the `Generating certificate` step above.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+Finally, click [uielement]#Edit method# to save the changes.
+
+image::samlv2/aiven/saml-configuration.png[Configuring SAML in Aiven.,width=1200]
+
+Go to the beginning of the page and click the [field]#Enable authentication method# switch to enable it. Now scroll down to [uielement]#Signup and link accounts URLs# and  copy the [uielement]#Signup URL#. You'll use it to test the integration and then share it with others to log into your Aiven application.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:111]:
```suggestion
Go to the beginning of the page and click the [field]#Enable authentication method# switch to enable logging in with FusionAuth.

Scroll down to [uielement]#Signup and link accounts URLs# and  copy the [uielement]#Signup URL#. You'll use it to test the integration and then share it with others to log into your Aiven application.
```

File: site/docs/v1/tech/samlv2/aiven.adoc [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+image::samlv2/aiven/copy-signup-url.png[Copy Signup URL from Aiven.,width=1200]
+
+== Testing
+
+Open the [uielement]#Signup URL# you copied above in an incognito tab or using another browser. You'll reach the signup page below.
+
+image::samlv2/aiven/login-sso.png[Logging in to Aiven.,width=542,role=custom]
+
+Click on [uielement]#Sign up with FusionAuth#, which will redirect you to the FusionAuth login page. Fill in your credentials and submit the form to be taken back to the Aiven console, already logged in and part of the `Developers` team.
```
Comment (mooreds) [inline site/docs/v1/tech/samlv2/aiven.adoc:121]: Do you need to register the user for the Aiven application?

Do you need to have any specific roles in FusionAuth that map to roles in Aiven?

**Review by mooreds — CHANGES_REQUESTED** [review]

File: site/docs/v1/tech/samlv2/aiven.adoc (hunk 13 above)
Comment (mooreds) [reply site/docs/v1/tech/samlv2/aiven.adoc:128]: @vcampitelli did you miss this?

Comment (vcampitelli) [reply site/docs/v1/tech/samlv2/aiven.adoc:138]: No, it was late and I decided to leave it for today, that's why I didn't re-request your review yesterday :smile: I just did it. 

And no, no specific roles are needed.

**Review by mooreds — APPROVED** [review]

**Pull request merged**